Fix the use of fsspec transactions

alanhdu opened the pull request:
fsspec transactions do not support concurrency and assumes that there is at most 1 running transaction per filesystem. This is *not* true in our usage, where because of multi-threading we usually have multiple concurrent transactions running at once.

Previously, this would just (unsafely) pass but lead to hard-to-debug race conditions (since the commit of one transaction will blow away the state of the other transaction). In fsspec 2024.3.0, trying to commit concurrent transactions will actually crash (see the code at https://github.com/fsspec/filesystem_spec/blob/76ca4a68885d572880ac6800f079738df562f02c/fsspec/transaction.py#L39 -- because each filesystem can have a single transaction, this tear-down logic will error).

Instead, let's manually handle committing / discarding changes to the file. This does this "the old-fashioned way" instead of using `fsspec`'s commit/rollback behavior because the internal PathManagerFileSystem used for `iopath` does not properly support that behavior.

I don't have a minimal test-case, but in Meta this solves a broken test on `fsspec >= 2024.3.0`:

Before: https://www.internalfb.com/intern/testinfra/testrun/7318349626774607
After: https://www.internalfb.com/intern/testinfra/testrun/2251800062722633

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @soulitzer @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn

alanhdu commented:
/easycla

Skylion007 commented:
Do we have an issue filed with fsspec, they would probably look into a fix for this (but this is a great temp workaround).

Skylion007 commented on torch/distributed/checkpoint/_fsspec_filesystem.py:
```diff
@@ -33,8 +34,16 @@ def create_stream(
+        # filesystem to isolate the transaction state. This is safe as
+        # long as we do not call `create_stream` concurrently on the
+        # same paths.
+        fs = copy.copy(self.fs)
```
Does this actually solve the problem or does it need to be a deep copy?

Skylion007 replied:
Not sure a shallow copy would be safe depending on how transactions are implemented on the backend for the specific fsspec endpoint.

Skylion007 replied:
On second thought, I think this is likely safe only if the file system wasn't already in transaction from a 3rd party lib etc. We could add an assert on that boolean, but that wouldn't fully protect from the race condition.

Skylion007 reviewed (commented):
Left some comments.

Skylion007 commented:
Looks like fixing the typing in this file also allowed flake8 and ruff to properly parse the types and find a few import issues.

alanhdu commented:
@Skylion007 : re the copy, I've updated the implementation to be (hopefully) a little less dependent on the exact implementation -- here, we directly use `fs._open` and handle the commit + rollback manually. This relies on a private part of the interface (`AbstractFileSystem._open`), but that might be better than relying on the internal details of how transactions are implemented wrt `copy`?

Let me know what you think -- I'm happy to remove the last commit if you think it'd be cleaner without it.

Skylion007 commented:
@martindurant Any thoughts on which approach is best?

martindurant commented:
It is true that fsspec transactions were not designed with parallelism in mind, nor are reentrant. It would take some effort to correct this.

Before getting into details, can someone comment on the purpose of transactions in general? Any file can be opened with autocommit=False, and then be cancelled or committed, so you can manage a container of these however you like.

alanhdu commented:
> Before getting into details, can someone comment on the purpose of transactions in general? Any file can be opened with autocommit=False, and then be cancelled or committed, so you can manage a container of these however you like.

I admit I don't fully understand the code, but my read from the initial PR (https://github.com/pytorch/pytorch/pull/112191) is just to ensure there aren't half-written files on some kind of failure. So we actually *need* any concurrency support.

The code I pushed does what you suggest, manually using `autocommit=False` + `commit`/`discard` (although there's a bit of ceremony involved it in actually getting it to work).

martindurant commented:
> The code I pushed does what you suggest

Perfect :). Then you shouldn't need to copy any filesystem instances.

alanhdu commented on torch/distributed/checkpoint/_fsspec_filesystem.py:
```diff
@@ -1,14 +1,13 @@
 # Mypy will not try inferring the types of any 3rd party libraries installed.
 # mypy: ignore-errors
+from __future__ import annotations
```
I thought I needed ti for the `self.fs: Optional[AbstractFileSystem]` annotation, but I guess not? Will change.

Skylion007 replied:
Nah, you just need to move AbstractFileSystem to the TYPE_CHECKING box. ruff should apply all these fixes for you if run `--fix`

martindurant replied:
I believe it allows `AbstractFileSystem | None` as an alternative in whichever python this wasn't yet a thing for (3.9?).

Skylion007 replied:
@martindurant That's orthogonal and doesn't play nicely with some other tools like legacy torch JIT. + It's just syntactic sugar.

alanhdu commented on torch/distributed/checkpoint/_fsspec_filesystem.py:
```diff
@@ -33,9 +36,29 @@ def create_stream(
+        #
+        # This is safe as long as you don't call `create_stream` on
+        # the same path concurrently
+        autocommit = "r" in mode
```
Unfortunately, running `stream.commit()` doesn't always work when things are set to read mode: for instance ,with a local file system, https://github.com/fsspec/filesystem_spec/blob/76ca4a68885d572880ac6800f079738df562f02c/fsspec/implementations/local.py#L368 special-cases the code such that you can't actually `commit` on read mode (since `self.temp` will not be declared).

I cold modify this to check that `mode != "rb"` I guess to be more exact? I agree that `r+` also exists...

martindurant replied:
There is indeed a mismatch here, due to most remote filesystems only supporting "w", "r" and maybe "a". Files should have a `.writable()` method, but of course that's only present after open.

alanhdu replied:
I think `.writable()` is exactly what we need here actually -- I can just set `autocommit=False` unconditionally and only run `.commit` or `.discard` if the stream is writable.

pradeepfn commented on torch/distributed/checkpoint/_fsspec_filesystem.py:
```diff
@@ -33,9 +35,28 @@ def create_stream(
+        # _open only supports binary mode
+        if "b" not in mode:
+            with self.create_stream(path, mode.replace("t", "") + "b") as stream:
+                yield io.TextIOWrapper(stream)
```
Can you help me what is happening here? 
It seems like we don't have to handle the same discard/commit semantics in binary mode (?). Can you help me understand why? ( or a link to appropriate reading). thanks.

alanhdu replied:
> It seems like we don't have to handle the same discard/commit semantics in binary mode (?).

That's not quite right. The basic issue is that `fs._open` is only guaranteed to be implemented for binary mode -- so for text mode we just (recursively) get the binary stream (which has the discard + commit semantics) and then wraps it in an `io.TextIOWrapper`. The recursion ensures that the binary stream also has the same commit + discard semantics.

martindurant replied:
This is right. It would have made sense for fsspec to wrap (heh) TextIOWrapper for "t" modes, to provide the same commit/discard methods.

pradeepfn commented:
During CP save, the file-system writer makes an atomic operation ( rename) on the checkpoint's .metadata file to mark the checkpoint as a valid checkpoint. 

Given FSSpecWriter uses the same logic, the atomicity/crash-consistency of the whole checkpoint is guaranteed by the .metadata atomic write.

Therefore, I don't really understand why we need atomic writes for individual write-items/tensors.

@fegin @LucasLLC  Can you help me understand. thanks.

Skylion007 commented:
> During CP save, the file-system writer makes an atomic operation ( rename) on the checkpoint's .metadata file to mark the checkpoint as a valid checkpoint. This happens in coordinator rank and is not a parallel operation.
> 
> Given FSSpecWriter uses the same logic, the atomicity/crash-consistency of the whole checkpoint is guaranteed by the .metadata atomic write.
> 
> Therefore, I don't understand why we need atomic writes for individual write-items/tensors ( which happens parallely and breaks the fs-spec TX as per the PR description).
> 
> @fegin @LucasLLC Can you help me understand. thanks.

Actually, rename is not guaranteed to be an atomic operation on all fsspec backends (it's a copy + delete on S3FS for example). It is atomic on some backends like ADLFS, but not on others. Fsspec.transaction is also not implemented for all backends either, btw.

alanhdu commented:
@Skylion007 @pradeepfn Do you have any suggestions on how to proceed here? Without this PR, our code is currently breaking with more recent fsspec, so I'm eager to try to get this merged.

If we think that the per-file crash-consistency is unnecessary, I'd be happy to just remove it entirely as well.

kwen2501 commented:
Ping @pradeepfn @fegin @LucasLLC

alanhdu commented:
Sorry to bug you again, but @pradeepfn @fegin @LucasLLC, but I'm interested in figuring out what the right path forward here is. I would like to be able to use newer versions of fsspec, which we currently cannot do with PyTorch right now.

Is there something else that needs to happen to get this fix merged?

alanhdu commented:
@pytorchbot merge -r

alanhdu commented:
@pytorchbot merge

alanhdu commented:
@pytorchbot merge

ZainRizvi commented:
@pytorchbot revert -c ghfirst -m "Breaking internally. See D65551490"

alanhdu committed:
Do not use a concurrent fsspec transaction

fsspec does not currently support concurrent transactions, and
more recent versions make this an error. Since we are only doing
single-file transactions, for now we can just manually try to
delete the file (if necessary) after errors.

alanhdu commented:
Sorry to bug you again, but @Skylion007 @pradeepfn @wz337 could I get another review? I have changed the implementation strategy (to be much simpler) so that things will work with internal fsspec implementations (like Manifold).

alanhdu commented:
Another ping @Skylion007 @pradeepfn @wz337 about the review. I am particularly interested about whether (at this point) we should just not have this per-file rollback attempt at all (since it wasn't even working before for filesystems that don't implement transactions, like Manifold).

alanhdu committed:
Fix lintrunner errors

alanhdu commented:
@pytorchbot merge
